sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala

```diff
@@ -17,53 +17,234 @@
+    case ByteType => PhysicalByteType
+    case ShortType => PhysicalShortType
+    case IntegerType => PhysicalIntegerType
+    case LongType => PhysicalLongType
+    case FloatType => PhysicalFloatType
+    case DoubleType => PhysicalDoubleType
+    case DecimalType.Fixed(p, s) => PhysicalDecimalType(p, s)
+    case BinaryType => PhysicalBinaryType
+    case _ => UninitializedPhysicalType
+  }
+}
+
+trait PhysicalPrimitiveType
+
+sealed abstract class TypedPhysicalDataType extends PhysicalDataType {
+  private[sql] type InternalType
+}
+
+sealed abstract class OrderedPhysicalDataType extends TypedPhysicalDataType {
+  private[sql] def ordering: Ordering[InternalType]
+}
+
+object OrderedPhysicalDataType {
+  def apply(dt: DataType): OrderedPhysicalDataType =
+    PhysicalDataType(dt).asInstanceOf[OrderedPhysicalDataType]
+
+  def ordering(dt: DataType): Ordering[Any] = {
+    try apply(dt).ordering.asInstanceOf[Ordering[Any]] catch {
+      case NonFatal(_) =>
+        throw QueryExecutionErrors.unsupportedTypeError(dt)
+    }
+  }
+}
+
+sealed abstract class PhysicalAtomicType extends OrderedPhysicalDataType {
+  private[sql] val tag: TypeTag[InternalType]
+}
+
+object PhysicalAtomicType extends PhysicalAtomicType {
+  def apply(dt: AtomicType): PhysicalAtomicType = dt match {
+    case ByteType => PhysicalByteType
+    case ShortType => PhysicalShortType
+    case IntegerType => PhysicalIntegerType
+    case LongType => PhysicalLongType
+    case StringType => PhysicalStringType
+    case FloatType => PhysicalFloatType
+    case DoubleType => PhysicalDoubleType
+    case DecimalType.Fixed(p, s) => PhysicalDecimalType(p, s)
+    case BooleanType => PhysicalBooleanType
+    case BinaryType => PhysicalBinaryType
+    case TimestampType => PhysicalLongType
+    case TimestampNTZType => PhysicalLongType
+    case DayTimeIntervalType(_, _) => PhysicalLongType
+    case DateType => PhysicalIntegerType
+    case _ => throw QueryExecutionErrors.unsupportedOperationExceptionError()
+  }
+
+  override private[sql] val tag = null
+  override private[sql] def ordering = null
+  override private[sql] type InternalType = Null
+}
+
+sealed abstract class PhysicalNumericType extends PhysicalAtomicType {
+}
+
+object PhysicalNumericType extends PhysicalNumericType {
+  def apply(nt: NumericType): PhysicalNumericType = {
+    PhysicalDataType(nt).asInstanceOf[PhysicalNumericType]
+  }
+
+  override private[sql] val tag = null
+  override private[sql] def ordering = null
+  override private[sql] type InternalType = Null
+}
+
+sealed abstract class PhysicalIntegralType extends PhysicalNumericType {
+}
+
+object PhysicalIntegralType extends PhysicalIntegralType {
+  def apply(dt: DataType): PhysicalIntegralType = dt match {
+    case IntegerType => PhysicalIntegerType
+    case ByteType => PhysicalByteType
+    case ShortType => PhysicalShortType
+    case LongType => PhysicalLongType
+    case _ => UninitializedPhysicalIntegralType
```

Comment:
This should just throw. Please do the same as for `OrderedPhysicalDataType `, and friends.